### PR TITLE
Added test cases for quadratic forms with matrix expressions

### LINKED ISSUE #9817: Quadratic forms with matrix expressions are handled poorly
There seems to be a number of issues with the handling of quadratic forms. 

``` python
Python 2.7.9 (default, Apr  2 2015, 15:33:21) 
Type "copyright", "credits" or "license" for more information.

IPython 2.3.0 -- An enhanced Interactive Python.
?         -> Introduction and overview of IPython's features.
%quickref -> Quick reference.
help      -> Python's own help system.
object?   -> Details about 'object', use 'object??' for extra details.

In [1]: import sympy as s

In [2]: n = 3

In [3]: v = s.MatrixSymbol('v',3,1)

In [4]: A = s.MatrixSymbol('A',3,3) 

In [5]: quadratic = v.T*A*v

In [6]: x = s.Matrix([i+1 for i in xrange(n)])

In [7]: X = s.Identity(n)

In [8]: subbed = quadratic.xreplace({v:x,A:X})

In [9]: print subbed
Matrix([
[1],
[2],
[3]])'*I*Matrix([
[1],
[2],
[3]])

In [10]: print(subbed.as_explicit())
Matrix([[14]])

In [11]: print(s.simplify(s.trace(subbed)))
Trace(Matrix([
[1],
[2],
[3]])'*I*Matrix([
[1],
[2],
[3]]))
```

Many equations use the quadratic form v^T_A_v and expect a scalar, which is not the case in sympy. Substitution with xreplace does not lead to a resolution of the form, but if this is forced with as_explicit we get a (1,1) matrix which conflicts with scalar symbols. Taking the trace does circumvent this, but apparently prevents evaluation of the internal expression. 
I also tried v.dot(A*v) which would be a reasonable compromise, but even when the dimension is fixed to (n,1) MatrixSymbol does not believe it can apply the .dot method. As soon as the vector is instantiated with vector value x it can call the method. 

Finally, putting the xreplace outside works, but simplify fails on the now supposedly scalar expression.

``` python
In [15]: s.simplify((s.trace(quadratic.as_explicit())).xreplace({v:x,A:X}))
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-15-2efc33ac8c3f> in <module>()
----> 1 s.simplify((s.trace(quadratic.as_explicit())).xreplace({v:x,A:X}))

/usr/local/lib/python2.7/dist-packages/sympy/simplify/simplify.pyc in simplify(expr, ratio, measure, fu)
   3706     _e = cancel(expr)
   3707     expr1 = shorter(_e, _mexpand(_e).cancel())  # issue 6829
-> 3708     expr2 = shorter(together(expr, deep=True), together(expr1, deep=True))
   3709 
   3710     if ratio is S.Infinity:

/usr/local/lib/python2.7/dist-packages/sympy/simplify/simplify.pyc in shorter(*choices)
   3700         if not has_variety(choices):
   3701             return choices[0]
-> 3702         return min(choices, key=measure)
   3703 
   3704     expr = bottom_up(expr, lambda w: w.normal())

/usr/local/lib/python2.7/dist-packages/sympy/core/function.pyc in count_ops(expr, visual)
   2250         while args:
   2251             a = args.pop()
-> 2252             if a.is_Rational:
   2253                 #-1/3 = NEG + DIV
   2254                 if a is not S.One:

AttributeError: 'str' object has no attribute 'is_Rational'
```

Possible solutions could be improved simplification of matrix expressions, a VectorSymbol class (or awareness of dot), or a quadratic form operator.

### PR DESCRIPTION
**Pull request by 0sidharth**
#### References to other Issues or PRs

Fixes #9817 

#### Brief description of what is fixed or changed
The issues pointed out in #9817 are already fixed on the current master. Tests for the same have been added here.

#### Release Notes

NO ENTRY

### CODE REVIEW

File: sympy/simplify/tests/test_simplify.py [hunk 1]
```diff
@@ -734,6 +734,16 @@ def test_issue_9324_simplify():
...
 
 
+def test_issue_9817_simplify():
+    from sympy.matrices.expressions import Identity, trace
+    v = MatrixSymbol('v', 3, 1)
+    A = MatrixSymbol('A', 3, 3)
+    x = Matrix([i + 1 for i in range(3)])
+    X = Identity(3)
+    quadratic = v.T * A * v
+    assert simplify((trace(quadratic.as_explicit())).xreplace({v:x, A:X})) == 14
```
Comment (oscarbenjamin) [inline sympy/simplify/tests/test_simplify.py:744]: What exactly is this test for?

Comment (0sidharth) [reply sympy/simplify/tests/test_simplify.py:744]: As reported in the original issue, there was an `AttributeError` being thrown on trying to execute this line, i.e. trying to simplify this scalar expression 
> Finally, putting the xreplace outside works, but simplify fails on the now supposedly scalar expression.

Comment (oscarbenjamin) [reply sympy/simplify/tests/test_simplify.py:744]: Can you add comments to these tests explaining what they are for?

Comment (0sidharth) [reply sympy/simplify/tests/test_simplify.py:744]: Sure will do, just one question. You mean adding comments that explain why these tests should pass right? Or should the comments explain what error was being thrown?

Comment (oscarbenjamin) [reply sympy/simplify/tests/test_simplify.py:744]: Explain what was not working before. It doesn't look as if the simplify call is doing anything so the purpose of the test needs to be explained.

**Comment by czgdp1807** [comment]
Will merge tomorrow if no objections raised. All comments seem to be addressed here.

**Comment by 0sidharth** [comment]
ping @czgdp1807

**Pull request merged**